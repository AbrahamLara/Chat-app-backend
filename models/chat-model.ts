import { Sequelize, UUID, UUIDV4 } from 'sequelize';
import { DataTypes, Models, SequelizeModel } from '../src/utils/database-utils';

export default function (
  sequelize: Sequelize,
  types: DataTypes
): typeof SequelizeModel {
  class Chat extends SequelizeModel {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static associate(models: Models) {
      Chat.belongsToMany(models.User, {
        through: models.UserChat,
        foreignKey: 'chatID',
      });
    }
  }

  Chat.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      isGroup: {
        type: types.BOOLEAN,
        allowNull: false,
      },
      name: {
        type: types.STRING,
        allowNull: false,
      },
    },
    {
      sequelize,
      modelName: 'Chat',
    }
  );
  return Chat;
}
