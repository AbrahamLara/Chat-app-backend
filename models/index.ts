import { DataTypes, Sequelize } from 'sequelize';
import User from './user';
import { Models } from '../utils/database-types';
// Get the development config from config.json
const config = require(`${__dirname}/../config/config.json`).development;
// Connect to the database.
const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  config
);

const models: Models = {
  User: User(sequelize, DataTypes),
};
// Begin to make associations between models if there is a model that requires it.
Object.keys(models).forEach(modelName => {
  const model = models[modelName];
  if (model.associate) {
    model.associate(models);
  }
});

export { sequelize, Sequelize, models };
