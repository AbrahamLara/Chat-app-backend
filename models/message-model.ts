import { Sequelize, UUID, UUIDV4 } from 'sequelize';
import { DataTypes, Models, SequelizeModel } from '../src/utils/database-utils';

export default function (
  sequelize: Sequelize,
  types: DataTypes
): typeof SequelizeModel {
  class Message extends SequelizeModel {
    /**
     * Helper method for defining associations.
     * This method is not a p art of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    static associate(models: Models) {
      // define association here
      Message.belongsTo(models.User, {
        foreignKey: 'userID',
        targetKey: 'id',
      });
      Message.belongsTo(models.Chat, {
        foreignKey: 'chatID',
        targetKey: 'id',
      });
      Message.belongsToMany(models.User, {
        through: models.MessageRecipient,
        foreignKey: 'messageID',
      });
    }
  }
  Message.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      message: {
        type: types.STRING,
        allowNull: false,
      },
      userID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Users',
          key: 'id',
        },
      },
      chatID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Chats',
          key: 'id',
        },
      },
    },
    {
      sequelize,
      modelName: 'Message',
    }
  );
  return Message;
}
