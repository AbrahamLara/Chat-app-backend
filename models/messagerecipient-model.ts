import { Sequelize, UUID, UUIDV4 } from 'sequelize';
import { Models, SequelizeModel } from '../src/utils/database-utils';

export default function (sequelize: Sequelize): typeof SequelizeModel {
  class MessageRecipient extends SequelizeModel {
    /**
     * Helper method for defining associations.
     * This method is not a part of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    static associate(models: Models) {
      MessageRecipient.belongsTo(models.User, {
        foreignKey: 'userID',
        targetKey: 'id',
      });
      MessageRecipient.belongsTo(models.UserChat, {
        foreignKey: 'userChatID',
        targetKey: 'id',
      });
      MessageRecipient.belongsTo(models.Message, {
        foreignKey: 'messageID',
        targetKey: 'id',
      });
    }
  }
  MessageRecipient.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      userID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Users',
          key: 'id',
        },
      },
      userChatID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'UserChats',
          key: 'id',
        },
      },
      messageID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Messages',
          key: 'id',
        },
      },
    },
    {
      sequelize,
      modelName: 'MessageRecipient',
    }
  );
  return MessageRecipient;
}
