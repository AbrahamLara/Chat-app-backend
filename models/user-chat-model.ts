import { Sequelize, UUID, UUIDV4 } from 'sequelize';
import { Models, SequelizeModel } from '../src/utils/database-utils';

export default function (sequelize: Sequelize): typeof SequelizeModel {
  class UserChat extends SequelizeModel {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static associate(models: Models) {
      UserChat.belongsTo(models.User, {
        foreignKey: 'userID',
        targetKey: 'id',
      });
      UserChat.belongsTo(models.Chat, {
        foreignKey: 'chatID',
        targetKey: 'id',
      });
      UserChat.hasMany(models.MessageRecipient, {
        foreignKey: 'userChatID',
        sourceKey: 'id',
      });
    }
  }
  UserChat.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      userID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Users',
          key: 'id',
        },
      },
      chatID: {
        type: UUID,
        allowNull: false,
        references: {
          model: 'Chats',
          key: 'id',
        },
      },
    },
    {
      sequelize,
      modelName: 'UserChat',
    }
  );
  return UserChat;
}
