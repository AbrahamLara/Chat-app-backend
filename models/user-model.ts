import { Sequelize, UUID, UUIDV4 } from 'sequelize';
import { DataTypes, Models, SequelizeModel } from '../src/utils/database-utils';

export default function (
  sequelize: Sequelize,
  types: DataTypes
): typeof SequelizeModel {
  class User extends SequelizeModel {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static associate(models: Models) {
      User.belongsToMany(models.Chat, {
        through: models.UserChat,
        foreignKey: 'userID',
      });
      User.hasMany(models.Message, {
        foreignKey: 'userID',
        sourceKey: 'id',
      });
      User.belongsToMany(models.Message, {
        through: models.MessageRecipient,
        foreignKey: 'userID',
      });
    }
  }

  User.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      name: {
        type: types.TEXT,
        allowNull: false,
      },
      email: {
        type: types.STRING,
        allowNull: false,
        unique: true,
      },
      password: {
        type: types.STRING,
        allowNull: false,
      },
    },
    {
      sequelize,
      modelName: 'User',
      indexes: [
        {
          name: 'users_name_trigram',
          using: 'GIN',
          concurrently: true,
          fields: [sequelize.Sequelize.literal('name gin_trgm_ops')],
        },
      ],
    }
  );
  return User;
}
