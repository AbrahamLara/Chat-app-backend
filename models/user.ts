import { Model, Sequelize, UUID, UUIDV4 } from 'sequelize';
import { DataTypes, Models, SequelizeModel } from '../utils/database-types';

export default function (
  sequelize: Sequelize,
  types: DataTypes
): typeof SequelizeModel {
  class User extends Model {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    static associate(models: Models) {
      // Defined model associations
    }
  }

  User.init(
    {
      id: {
        type: UUID,
        primaryKey: true,
        defaultValue: UUIDV4,
      },
      firstName: {
        type: types.STRING,
        allowNull: false,
      },
      lastName: {
        type: types.STRING,
        allowNull: false,
      },
      email: {
        type: types.STRING,
        allowNull: false,
        unique: true,
      },
      password: {
        type: types.STRING,
        allowNull: false,
      },
    },
    {
      sequelize,
      modelName: 'User',
    }
  );
  return User;
}
