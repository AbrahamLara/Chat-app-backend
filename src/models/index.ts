import { DataTypes, Sequelize } from 'sequelize';
import User from './user-model';
import { Models } from '../utils/database-utils';

const NODE_ENV = process.env.NODE_ENV || 'development';

// Get the env config from config.json
const config = require('../../config/config.json')[NODE_ENV];
// Connect to the database.
const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  config
);

const models: Models = {
  User: User(sequelize, DataTypes),
};
// Begin to make associations between models if there is a model that requires it.
Object.keys(models).forEach(modelName => {
  const model = models[modelName];
  if (model.associate) {
    model.associate(models);
  }
});

export { sequelize, Sequelize, models };
