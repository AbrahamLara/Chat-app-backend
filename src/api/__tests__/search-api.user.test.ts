import { Server } from 'http';
import { SuperAgentTest } from 'supertest';
import {
  MOCK_JWT,
  MOCK_USER,
  mockServerAndAgent,
} from '../../test_utils/mock-utils';
import mockUsers from '../../test_utils/sample_data/mock-users.json';
import { models } from '../../../models';
import { mockTokenUtils } from '../../test_utils/file_mocks/mock-token-utils';

const { User } = models;
const MOCK_SEARCH_USER_IDS = mockUsers.map(user => user.id);
let server: Server;
let agent: SuperAgentTest;

mockTokenUtils();

describe('Search user endpoint', () => {
  beforeEach(
    mockServerAndAgent((mockServer, testAgent) => {
      server = mockServer;
      agent = testAgent;
    })
  );

  afterEach(cb => {
    server.close(cb);
  });

  it('does not perform search without auth token', async () => {
    const response = await searchUserWithName(agent, 'e', '');
    // Expect the call to have failed because no token was provided.
    expect(response.status).toBe(401);
    expect(response.ok).toBeFalsy();
    expect(response.body).toMatchSnapshot();
  });

  it('does not return the name of the user who performed the search', async () => {
    // If an error occurs in the middle of the test, we should remove users that were already created  before attempting
    // to bulk create again.
    await removeMockUsers();
    await User.bulkCreate(mockUsers);

    const response = await searchUserWithName(agent, 'e', MOCK_JWT);

    // Expect the call to be successful and contain 10 user results and none of them being the mock user.
    expect(response.status).toBe(200);
    expect(response.ok).toBeTruthy();
    expect(response.body.results.length).toBe(10);
    expect(response.body.results).not.toContain({ name: MOCK_USER.name });

    await removeMockUsers();
  });
});

/**
 * Performs a user search from the search endpoint using the provided auth token.
 */
async function searchUserWithName(
  testAgent: SuperAgentTest,
  name: string,
  token: string
) {
  const authorizationValue = token ? `Bearer ${token}` : '';
  return testAgent
    .get(`/api/search/user?name=${name}`)
    .set('Authorization', authorizationValue);
}

/**
 * Remove every mock user that was inserted into the test database for search purposes.
 */
async function removeMockUsers() {
  await User.destroy({ where: { id: MOCK_SEARCH_USER_IDS } });
}
