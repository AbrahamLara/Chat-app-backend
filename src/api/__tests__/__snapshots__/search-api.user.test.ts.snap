// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search user endpoint does not perform search without auth token 1`] = `
Object {
  "message": "You are not authorized to perform this action",
}
`;
